keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+
+Users of a ResourceClaim don't need to be Pods. This KEP specifically supports
+Pods as users and describes how kube-scheduler and kubelet will deal with Pods
+that depend on a ResourceClaim, but the API and some custom resource driver
+might also be useful for controllers to manage resources without using those
+resources for Pods.
+
+### User Stories
+
+#### Cluster add-on development
+
+As a hardware vendor, I want to make my hardware available also to applications
+that run in a container under Kubernetes. I want to make it easy for a cluster
+administrator to configure a cluster where some nodes have this hardware.
+
+I develop two components, one that runs as part of the Kubernetes control plane
+and one that runs on each node, and package those inside container images. YAML
+files describe how to deploy my software on a Kubernetes cluster that supports
+dynamic resource allocation.
+
+Documentation for administrators explains how the nodes need to be set
+up. Documentation for users explains which parameters control the behavior of
+my hardware and how to use it inside a container.
+
+#### Cluster configuration
+
+As a cluster administrator, I want to make GPUs from vendor ACME available to users
+of that cluster. I prepare the nodes and deploy the vendor's components with
+`kubectl create`.
+
+I create a ResourceClass for the hardware with parameters that only I as the
+administrator am allowed to choose, like for example running a command with
+root privileges that does some cluster-specific initialization for each allocation:
+```
+apiVersion: cdi.k8s.io/v1alpha1
+metadata:
+  name: acme-gpu
+driverName: gpu.acme.com
+parameters:
+  initCommand:
+  - /usr/local/bin/acme-gpu-init
+  - --cluster
+  - my-cluster
+```
+
+#### Partial GPU allocation
+
+As a user, I want to use a GPU as accelerator, but don't need exclusive access
+to that GPU. Running my workload with just 2Gb of memory is sufficient. This is
+supported by the ACME GPU hardware. I know that the administrator has created
+an "acme-gpu" ResourceClass.
+
+For a simple trial, I create a Pod directly where two containers share the same subset
+of the GPU:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: device-consumer
+spec:
+  resourceClaims:
+  - name: "gpu" # this name gets referenced below under "claims"
+    template:
+      resourceClassName: "acme-gpu"
+      parameters:
+        memory: "2Gi"
+  containers:
+  - name: workload
+    image: my-app
+    command: ["/bin/program"]
+    resources:
+      requests:
+        memory: "64Mi"
+        cpu: "250m"
+      limits:
+        memory: "128Mi"
+        cpu: "500m"
+      claims:
+        - "gpu"
+  - name: monitor
+    image: my-app
+    command: ["/bin/other-program"]
+    resources:
+      requests:
+        memory: "32Mi"
+        cpu: "25m"
+      limits:
+        memory: "64Mi"
+        cpu: "50m"
+      claims:
+      - "gpu"
+```
+
+This request triggers resource allocation on a node that has a GPU device with
```

Comment:
this device triggers the allocation?  is it the node local driver?  is the driver responsible for listing and watching all pods?  are the pods filtered by the local node name?  is it watching all resource claims independent of a pod? both?

Reply:
Did the text below answer these questions?

The user story was just meant to show how a users uses the API, not explain what happens under the hood.

Reply:
my question was trying to figure out scale and qps concerns.

- how many resources will each kubelet in a cluster watch
- how many resources will each kubelet write status against
- are those resources filtered by some security dimension (similar to how kubelets can only see secrets for associated pods bound to its node)

Reply:
These are valid questions, it just wasn't clear to me whether you wanted to have them addressed by the KEP in this section.

I need to think more about how kubelet will watch ResourceClaims and then add more about that in the design section below.

Reply:
I've revised the API such that kubelet can filter ResourceClaims by a field that assigns claims to the node(s) on which is going to be used by a pod.

The effect is that each kubelet instance only needs to receive and update ResourceClaim objects that are actually in use on the node.

There is nothing that stops kubelet from requesting other claims and modifying them, so this is just addressing scalability. It's not a security mechanism that protects against a compromised or faulty kubelet.

Can kubelet write all Pod statuses? If not, can someone point me towards the code that implements that protection?

Reply:
Kubelet can only write status on pods on its own node, I think, but I don't know where that logic lives.